js/spreadsheet.js

```diff
@@ -68,244 +68,196 @@ var spreadsheet = function(options) {
-      // When we got a null value in the cell it means that we reached the range borders
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][firstCol] === null) {
-          lastRow = i;
+    if (currentIndex || currentIndex === 0) {
+      for (currentIndex; currentIndex < tableData.length; currentIndex++) {
+        if (tableData[currentIndex][colIndex] === null) {
           break;
         }
       }
-
-      lastRow = _.max([lastRow - 1, 0]);
-    } else if (moveTo.right) {
-      // When data located on the right of the selected cell
-      firstCol = startAt[1];
-
-      for (var i = firstCol + 1; i < allData.length; i++) {
-        if (allData[startAt[0]][i] === null) {
-          lastCol = i - 1;
+    } else {
+      for (currentIndex; currentIndex < tableData.length; currentIndex++) {
+        if (tableData[currentIndex] === null) {
           break;
         }
       }
+    }
 
-      for (var i = startAt[0]; i > 0; i--) {
-        if (allData[i][lastCol] === null) {
-          firstRow = i ? i - 1 : i;
-        }
-      }
+    return currentIndex;
+  }
 
-      firstRow = firstRow || 0;
+  /**
+   * Function to get coordinates which we need to select
+   * 
+   * @param {Array} cellPosition - an array of the selected coordinates
+   * 
+   * @returns {Array/Boolean} - coordinates that need to be selected. Example of the returned data: [[startRow, startCol, endRow, endCol]]
+   */
+  function coordinatesToSelect(cellPosition) {
+    var firstCol, lastCol, firstRow, lastRow, tableData;
+
+    // Returns array of the data from the table with Handsontable API
+    tableData = hot.getData();
+    selectedCellPosition = {
+      row: cellPosition[0][0],
+      col: cellPosition[0][1]
+    };
 
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][lastCol] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.top) {
-      // When data located on the top of the selected cell
-      lastRow = startAt[0];
+    var nearestDataAt = nearestCellsWithData(selectedCellPosition, tableData);
 
-      for (var i = lastRow - 1; i > 0; i--) {
-        if (allData[i][startAt[1]] === null) {
-          firstRow = i;
-          break;
-        }
-      }
+    // Return false in the case when we need to select all table data
+    if (nearestDataAt.selectAll) {
+      return false;
+    }
 
-      firstRow = firstRow || 0;
+    if (nearestDataAt.left) {
+      var col = selectedCellPosition.col - 1;
 
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[firstRow][i] === null) {
-          firstCol = i ? i + 1 : i;
-          break;
-        }
-      }
+      firstCol = getStartIndex(tableData[selectedCellPosition.row], col);
 
-      firstCol = firstCol || 0;
+      // If column index is not 0 it means that data what we need to select is in the next column
+      firstCol = firstCol === 0 ? firstCol : firstCol + 1;
```

Comment:
If I'm reading this right, if `firstCol` is `1`, you'd end up adding `1` to it and make it `2`? What's the reason behind doing this?

Reply:
@tonytlwu it would be easier to explain on an example:
How selection will behave without this line.
https://share.getcloudapp.com/yAu2Pode

And with this line.
https://share.getcloudapp.com/WnuG1woX

This behavior occurs because of how we searching for the coordinates. 
Our functions `getStartIndex` and `getEndIndex` are looking for the `null` in value of the cell to determine that there is no data in that cell. That is why they are returning an index of the cell with null, but we are interested to select only cells with data, that is why we adding to the row or cell 1 here.